homeassistant/components/avreceiver/config_flow.py

```diff
@@ -0,0 +1,128 @@
+"""Config flow for AV Receiver."""
+from urllib.parse import urlparse
+
+from pyavreceiver import factory
+from pyavreceiver.error import AVReceiverIncompatibleDeviceError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import ssdp
+from homeassistant.const import CONF_HOST, CONF_ID
+from homeassistant.core import callback
+
+from .const import (
+    CONF_ZONE2,
+    CONF_ZONE3,
+    CONF_ZONE4,
+    DATA_DISCOVERED_HOSTS,
+    DEFAULT_ZONE_DISABLED,
+    DOMAIN,
+)
+
+
+def format_title(host: str) -> str:
+    """Format the title for config entries."""
+    return f"AV Receiver ({host})"
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class AVReceiverFlowHandler(config_entries.ConfigFlow):
+    """Define the config flow for AV Receiver."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry) -> "OptionsFlowHandler":
+        """Get the options flow."""
+        return OptionsFlowHandler(config_entry)
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a discovered Heos device."""
+        # Store discovered host
+        hostname = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION]).hostname
+        friendly_name = f"{discovery_info[ssdp.ATTR_UPNP_FRIENDLY_NAME]} ({hostname})"
+        serial = discovery_info[ssdp.ATTR_UPNP_SERIAL]
+        self.hass.data.setdefault(DATA_DISCOVERED_HOSTS, {})
+        self.hass.data[DATA_DISCOVERED_HOSTS][friendly_name] = hostname
+        # Abort if other flows in progress or an entry already exists
+        if self._async_in_progress() or self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        await self.async_set_unique_id(f"{DOMAIN}-{serial}")
+        # Show selection form
+        return self.async_show_form(step_id="user")
+
+    async def async_step_user(self, user_input=None):
+        """Get host from user and check connection."""
+        self.hass.data.setdefault(DATA_DISCOVERED_HOSTS, {})
+        # Try connecting to host
+        errors = {}
+        host = None
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            try:
+                avreceiver = await factory(host)
+                await avreceiver.init()
+                await avreceiver.disconnect()
+                self.hass.data.pop(DATA_DISCOVERED_HOSTS)
+                unique_id = (
+                    f"{DOMAIN}-{avreceiver.serial_number or avreceiver.mac or host}"
+                )
+                await self.async_set_unique_id(unique_id, raise_on_progress=False)
```

Comment:
It's modeled from the pyheos integration.  The reason given is "raise_on_progress is False here in case ssdp discovers heos first which would block the import" - however, this is in the import step which I removed.  I will look into it.

Looks like I need to add MAC to the ssdp integration...
https://github.com/home-assistant/core/blob/82746616fa92abc89a3a2b2e692c50ed4df2f97e/homeassistant/components/ssdp/__init__.py#L19-L36